### PR TITLE
Act refactor: require provider/refactoring/file and render contract

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Implements the canonical, operator-facing requirements for `weaver act refactor` (2.3.4): require provider, refactoring, and file flags, plus a deterministic, full-contract error when arguments are missing. Adds an ExecPlan living document to guide implementation, testing, and docs.
- Introduces a central requirements helper to expose valid providers, refactorings, and the required argument set, ensuring single-source truth across validation, tests, and user-facing guidance.
- Updates the argument parsing and routing flow to validate the full contract before runtime resolution and plugin execution.
- Extends tests (unit, contract, and feature) to cover the complete missing-arguments path, unsupported values, and explicit-provider scenarios.
- Updates docs and roadmaps to reflect the final MVP contract and validation gates.

## Changes
- Added new file: crates/weaverd/src/dispatch/act/refactor/requirements.rs
- Updated parsing/validation to require all three flags and to validate provider/refactoring against a canonical list
- Made provider a required field in RefactorArgs and tightened the build/validation flow
- Updated `crates/weaverd/src/dispatch/act/refactor/mod.rs` to reflect the new required-arguments contract in the command handler and logging
- Expanded tests:
  - Updated unit tests in crates/weaverd/src/dispatch/act/refactor/arguments.rs
  - Extended contract tests (crates/weaverd/src/dispatch/act/refactor/contract_tests.rs)
  - Extended refactor feature tests (crates/weaverd/tests/features/refactor.feature)
  - Amended router tests to reflect the new error surface for missing required flags
- Added ExecPlan doc: docs/execplans/2-3-4-return-argument-requirements-for-act-refactor.md
- Documentation updates:
  - docs/weaver-design.md updated to describe the MVP contract with explicit provider/refactoring validation
  - docs/users-guide.md updated to reflect the required flags and valid values
  - docs/roadmap.md updated to mark 2.3.4 as done where applicable

## Rationale
- The roadmap item 2.3.4 calls for surfacing the full operator contract when required arguments are missing, including the complete set of required flags and valid values. This change centralizes required flags, valid providers, and supported refactorings to a single source of truth and renders a deterministic, actionable error in one response, aligning with the Level 10 error-template policy and ensuring consistent behavior across parser, router, and runtime paths.
- By enforcing the contract before plugin resolution, file I/O, or backend startup, we reduce ambiguity for users and testers and keep the existing safety pipeline intact.

## Review guidance
- Verify that the ExecPlan communicates the complete contract, acceptance criteria, and staged approach.
- Ensure the new canonical requirements helper aligns with validation code, tests, and docs.
- Confirm that the updated error messages and guidance lines reflect the canonical provider/refactoring lists and the next-step command.
- Check alignment with related docs (weaver-design.md, users-guide.md, roadmap.md) as they are updated.

## Implementation plan (summary)
- Stage A: lock in required behavior with failing tests (missing-arguments path surfaces full contract)
- Stage B: introduce canonical requirements model (providers/refactorings/flags/public messages)
- Stage C: enforce the contract in parsing and routing (reject incomplete requests with single, actionable error)
- Stage D: update design and user docs to reflect final MVP contract
- Stage E: validate end-to-end (docs validators, Rust gates, tests)

## Validation plan
- Red-phase unit tests asserting the full missing-arguments contract is rendered
- Behavioural tests validating happy path, missing-arguments, and edge cases (partial flags, invalid values)
- Optional end-to-end CLI assertion to ensure operator-facing output matches acceptance criteria

## Docs impact
- ExecPlan doc added for 2.3.4
- Design and user docs updated to reflect the new contract and values
- Roadmap item 2.3.4 marked appropriately after verification

## Progress tracking
- This PR introduces the ExecPlan living document and implements the complete 2.3.4 argument contract for act refactor. Ongoing progress and status updates will be reflected in the docs as work proceeds.

---
◳ Generated by [DevBoxer](https://www.devboxer.com)

ℹ️ Tag @devboxerhub to address PR feedback

📎 **Task**: https://www.devboxer.com/task/c84a1800-74c2-4659-a655-422bcd862e05

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaverd/src/dispatch/act/refactor/requirements.rs](https://github.com/leynos/weaver/pull/102/files/7d339bec5034385175f7e4513616ed9c6bf31283#diff-83eb50c50a6684e20731d2f65524d4a3b4c6c5d2cc9880424b239bf446cca52f)

Comment on file

~~~rust
pub(crate) fn validate_provider(provider: &str) -> Result<(), DispatchError> {
    if supported_provider_names().contains(&provider) {
        Ok(())
    } else {
        Err(DispatchError::invalid_arguments(format!(
            "act refactor does not support provider '{provider}'\n\n{}",
            guidance_lines()
        )))
    }
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=102&biomarker=Code+Duplication&filename=crates%2Fweaverd%2Fsrc%2Fdispatch%2Fact%2Frefactor%2Frequirements.rs&method=)
The module contains 2 functions with similar structure: validate_provider,validate_refactoring

File: crates/weaverd/src/dispatch/act/refactor/refactor_helpers.rs [hunk 1]
```diff
@@ -59,6 +59,8 @@ pub(super) fn build_backends(socket_path: &Path) -> FusionBackends<SemanticBacke
 
 pub(super) fn standard_rename_args(file: &str) -> Vec<String> {
     vec![
+        String::from("--provider"),
+        String::from(provider_for_file(file)),
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaverd/src/dispatch/act/refactor/refactor_helpers.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### docs/users-guide.md (1)

`881-929`: _⚠️ Potential issue_ | _🟡 Minor_
**Remove the omitted-provider routing guidance later in the guide.**
This section makes `--provider` mandatory, but the later plugin-capabilities

section still says the daemon selects `rope` or `rust-analyzer` when

`--provider` is omitted. Delete or rewrite that later text, otherwise the guide

documents two incompatible operator flows.
As per coding guidelines "Ensure that any API or behavioural changes are reflected in the documentation in `docs/`".

##### 🤖 Prompt for AI Agents

File: crates/weaverd/tests/features/refactor.feature [hunk 2]
```diff
@@ -58,3 +58,13 @@ Feature: Act refactor
...
+
+  Scenario: Refactor reports the complete required argument contract
+    Given an act refactor request without the required flags
+    When the act refactor command executes
+    Then the refactor command fails with status 1
+    And the dispatch error contains "--provider <plugin>"
+    And the dispatch error contains "--refactoring <operation>"
+    And the dispatch error contains "--file <path>"
+    And the dispatch error contains "Providers: rope, rust-analyzer"
+    And the dispatch error contains "Refactorings: rename"
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaverd/tests/features/refactor.feature]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[crates/weaverd/src/dispatch/act/refactor/arguments.rs](https://github.com/leynos/weaver/pull/102/files/9523781ee68fc7cdb8dd4581de117e7f91efeeff#diff-66c26e26b394ed5e0e83ad7e903fc661bed7fd3d8cefa8d9bec5a3aa19b97818)

Comment on file

~~~rust

    #[test]
    fn missing_file_is_rejected() {
    fn missing_flag_value_is_rejected() {
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=102&biomarker=Code+Duplication&filename=crates%2Fweaverd%2Fsrc%2Fdispatch%2Fact%2Frefactor%2Farguments.rs&method=)
The module contains 6 functions with similar structure: tests.flag_as_value_is_rejected,tests.malformed_trailing_arguments_are_rejected,tests.missing_flag_value_is_rejected,tests.unexpected_top_level_flag_is_rejected and 2 more functions

**Comment by leynos** [comment]
@coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

~~~
Please address the comments from this code review:

## Overall Comments
- The canonical `SUPPORTED_PROVIDERS`/`SUPPORTED_REFACTORINGS` lists in `requirements.rs` are still manually duplicated; consider deriving them from the same manifest/runtime source that registers providers/refactorings so they can't silently drift when new capabilities are added.
- In `refactor_helpers::provider_for_file`, defaulting unknown extensions to `"rope"` can hide misconfigured tests or new-language scenarios; consider restricting this helper to known extensions and failing loudly (or requiring explicit provider in tests) when the mapping is incomplete.

## Individual Comments

### Comment 1
<location path="docs/execplans/2-3-4-return-argument-requirements-for-act-refactor.md" line_range="15" />
<code_context>
+
+## Purpose / big picture
+
+Roadmap item `2.3.4` closes UI gap `5b` by making `weaver act refactor` surface
+the full operator contract when required arguments are missing. Today the
+daemon stops at the first missing field and returns only one message, for
</code_context>
<issue_to_address>
**suggestion (review_instructions):** UI is introduced without being expanded; please define it on first use (for example, “user interface (UI) gap”).

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 2
<location path="docs/execplans/2-3-4-return-argument-requirements-for-act-refactor.md" line_range="57" />
<code_context>
+- Add both unit coverage and behavioural coverage using `rstest-bdd` v0.5.0.
+  Reuse the existing `act refactor` feature harness in `crates/weaverd/` rather
+  than creating a second bespoke BDD harness.
+- Keep the JSONL transport envelope unchanged. This task may improve the
+  `InvalidArguments` message body, but it must not add a new top-level response
+  type or widen the `stream` / `exit` protocol contract.
</code_context>
<issue_to_address>
**suggestion (review_instructions):** JSONL is introduced here without expansion; please define it on first use (for example, “JSON Lines (JSONL) transport envelope”).

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 3
<location path="docs/execplans/2-3-4-return-argument-requirements-for-act-refactor.md" line_range="269" />
<code_context>
+contract.
+
+`crates/weaver-e2e/tests/refactor_rope_cli_snapshots.rs` and
+`crates/weaver-e2e/tests/refactor_rust_analyzer_cli_snapshots.rs` exercise the
+operator-facing CLI workflows. A small end-to-end assertion or snapshot should
+be added here if needed to prove that the human-visible `weaver act refactor`
</code_context>
<issue_to_address>
**suggestion (review_instructions):** CLI is used for the first time in this section in “CLI workflows” without definition; please introduce it as “command-line interface (CLI)” on first use.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 4
<location path="docs/execplans/2-3-4-return-argument-requirements-for-act-refactor.md" line_range="174" />
<code_context>
+
+- The current missing-argument logic lives in the parser builder, not in the
+  handler. That is the right seam for this work because it lets the daemon
+  reject incomplete requests before plugin resolution, file I/O, or backend
+  startup.
+
</code_context>
<issue_to_address>
**suggestion (review_instructions):** I/O is introduced here without expansion; please define it as “input/output (I/O)” the first time it appears.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Developer Documentation | ⚠️ Warning | Pull request introduces undocumented requirements.rs module establishing canonical API for act refactor contract with seven public functions and outdated code examples in developers guide. | Add requirements.rs module documentation to developers-guide.md detailing seven public functions and their purpose, then update all refactor code examples to reflect SelectionMode::ExplicitProvider pattern. |
| Module-Level Documentation | ⚠️ Warning | Two new public functions introduced lack required documentation: `built_in_provider_names()` in manifests.rs and `standard_rename_args_for_provider()` in refactor\_helpers.rs missing docstrings, whilst requirements.rs correctly documents all public functions. | Add docstrings to both functions documenting their purpose and return values according to Rust documentation standards for public APIs. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

## Code Duplication
crates/weaverd/src/dispatch/act/refactor/arguments.rs: 

### What lead to degradation?
The module contains 6 functions with similar structure: tests.flag_as_value_is_rejected,tests.malformed_trailing_arguments_are_rejected,tests.missing_flag_value_is_rejected,tests.unexpected_top_level_flag_is_rejected and 2 more functions

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

File: crates/weaverd/src/dispatch/router.rs [hunk 3]
```diff
@@ -138,14 +138,25 @@ impl std::fmt::Debug for DomainRouter {
...
     /// Creates a new domain router with the workspace root.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `workspace_root` is not an absolute path. Dispatch handlers
+    /// resolve workspace-relative files against this root and rely on the
+    /// resulting paths to construct canonical file URIs.
     pub fn new(workspace_root: PathBuf) -> Self {
+        assert_absolute_workspace_root(workspace_root.as_path());
         Self {
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaverd/src/dispatch/router.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### docs/developers-guide.md (1)

`500-507`: _⚠️ Potential issue_ | _🟡 Minor_
**Align the explicit-provider example end-to-end.**
Replace `standard_rename_args("notes.py")` with

`standard_rename_args_for_provider("notes.py", "rope")` so the request example

matches `SelectionMode::ExplicitProvider` and

`requested_provider: Some("rope")`.

##### Suggested doc patch

```diff
-let request = command_request(standard_rename_args("notes.py"));
+let request = command_request(standard_rename_args_for_provider("notes.py", "rope"));
```

Triage: `[type:docstyle]`

##### 🤖 Prompt for AI Agents

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[crates/weaverd/src/dispatch/act/refactor/arguments.rs](https://github.com/leynos/weaver/pull/102/files/367baa2d812d12ea6d688b0fb9cda48d09a37e5d#diff-66c26e26b394ed5e0e83ad7e903fc661bed7fd3d8cefa8d9bec5a3aa19b97818)

Comment on file

~~~rust

    #[test]
    fn missing_file_is_rejected() {
    fn missing_flag_value_is_rejected() {
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=102&biomarker=Code+Duplication&filename=crates%2Fweaverd%2Fsrc%2Fdispatch%2Fact%2Frefactor%2Farguments.rs&method=)
The module contains 6 functions with similar structure: tests.flag_as_value_is_rejected,tests.malformed_trailing_arguments_are_rejected,tests.missing_flag_value_is_rejected,tests.unexpected_top_level_flag_is_rejected and 2 more functions

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
crates/weaverd/src/dispatch/act/refactor/arguments.rs: 

### What lead to degradation?
The module contains 6 functions with similar structure: tests.flag_as_value_is_rejected,tests.malformed_trailing_arguments_are_rejected,tests.missing_flag_value_is_rejected,tests.unexpected_top_level_flag_is_rejected and 2 more functions

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
crates/weaverd/src/dispatch/act/refactor/arguments.rs: 

### What lead to degradation?
The module contains 6 functions with similar structure: tests.flag_as_value_is_rejected,tests.malformed_trailing_arguments_are_rejected,tests.missing_flag_value_is_rejected,tests.unexpected_top_level_flag_is_rejected and 2 more functions

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

File: crates/weaverd/src/dispatch/act/refactor/mod.rs [hunk 4]
```diff
@@ -180,29 +182,25 @@ fn prepare_plugin_request(
...
-    let effective_operation = match args.refactoring.as_str() {
-        "rename" => {
-            apply_rename_symbol_mapping(&mut plugin_args, &args.file)?;
-            String::from("rename-symbol")
-        }
-        _ => args.refactoring.clone(),
-    };
+    let effective_operation = effective_operation(&args.refactoring)?;
+    if effective_operation == "rename-symbol" {
+        apply_rename_symbol_mapping(&mut plugin_args, file_path.as_path())?;
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaverd/src/dispatch/act/refactor/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
crates/weaverd/src/dispatch/act/refactor/arguments.rs: 

### What lead to degradation?
The module contains 6 functions with similar structure: tests.flag_as_value_is_rejected,tests.malformed_trailing_arguments_are_rejected,tests.missing_flag_value_is_rejected,tests.unexpected_top_level_flag_is_rejected and 2 more functions

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue.

### ❌ Failed checks (2 errors, 6 warnings, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Module-Level Documentation | ❌ Error | The public function `standard_rename_args_for_provider()` lacks required documentation with no `///` docstring explaining its purpose or parameters. | Add a docstring to `standard_rename_args_for_provider()` immediately before the function signature explaining its purpose and parameters. |
| Unit Architecture | ❌ Error | Three explicit review comments requesting architectural refinements remain unresolved: hard-coded sentinel duplicates contract logic in requirements.rs, test helper uses panic instead of Result propagation, and test\_support duplicates error-unwrapping logic instead of utilising shared helper. | Apply three refactoring diffs: modify map\_effective\_operation() to return Result<(String, CapabilityId), DispatchError>, convert then\_dispatch\_error\_contains() to use ok\_or\_else() instead of panic, replace inline match in test\_support with required\_result() helper. |
| User-Facing Documentation | ⚠️ Warning | PR introduces breaking change to CLI interface by making `--provider` mandatory, but lacks a dedicated migration document per repository patterns. | Create a migration guide following established conventions (e.g., `weaver-v0-1-1-migration-guide.md`) with breaking changes table, before/after examples, and rationale for the change. |
| Developer Documentation | ⚠️ Warning | The new pub(crate) function `standard_rename_args_for_provider()` lacks a docstring and is not explicitly documented in the developers-guide. | Add a `///` docstring to `standard_rename_args_for_provider()` and update `docs/developers-guide.md` to explicitly list it alongside `standard_rename_args()`. |
| Testing (Unit) | ⚠️ Warning | Three unaddressed review comments remain: panic in behaviour.rs violates Result error propagation, redundant capability\_for\_operation call in request\_building.rs, and inconsistent error-handling in test\_support/mod.rs. | Apply three pending refactors: replace panic with .ok\_or\_else, refactor map\_effective\_operation to return (String, CapabilityId), and consolidate error-unwrapping logic using required\_result helper. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | PR introduces deterministic error messages forming a structured operator contract, but tests use only substring assertions without snapshot testing to validate the complete UI contract structure and formatting stability. | Add snapshot tests using insta to validate complete error message output for each failure path, capturing required flags, provider/refactoring lists, and next-command guidance with full contract structure. |
| Domain Architecture | ⚠️ Warning | Pull request violates domain architecture by maintaining a hardcoded "rename-symbol" string sentinel in request\_building.rs instead of deriving operation-specific behaviour from the resolved CapabilityId. | Refactor map\_effective\_operation to return both the operation string and its resolved CapabilityId, obtained via capability\_for\_operation. Match on the resolved CapabilityId (not the string) to decide whether to apply rename-symbol-specific argument transformations. Additionally, replace the inline match in test\_support/mod.rs at lines 81–85 with the existing required\_result helper to eliminate duplication. |
| Observability | ⚠️ Warning | PR introduces operational validation gate without observability infrastructure: no metrics for failure categorisation, no structured logging at validation boundaries, and no tracing spans around validation flow. | Add structured logging at validation decision point, introduce metrics for failure types (missing-provider, unsupported-provider, unsupported-refactoring, malformed-arguments), and implement trace spans around validation flow. |
| Testing (Overall) | ❓ Inconclusive | Insufficient evidence to determine if all introduced behaviours are rigorously tested across multiple modules and validation paths. | Review test implementations in requirements.rs, arguments.rs, and feature tests, verifying assertions depend on actual validation logic. Add targeted negative–positive tests that would fail if validation were bypassed. |

**Comment by leynos** [comment]
@coderabbitai is the new ubicloud usage in CI documented in `docs/developer-guide.md`?

**Comment by leynos** [comment]
@coderabbitai please create new GitHub issues for the following:

### Out-of-scope items — proposed GitHub issues

Would you like me to open GitHub issues for the three out-of-scope warnings below?

1. **User-Facing Documentation**: Migration guide for the `--provider` breaking change (following the `docs/weaver-lsp-host-v0-1-0-migration-guide.md` convention).
2. **Testing (Compile-Time/UI)**: Add `insta` snapshot tests for the complete `InvalidArguments` error-message contract.
3. **Observability**: Structured logging at validation decision points, failure-type metrics, and tracing spans around the validation flow.
